File: 129_sum_root_to_leaf_numbers/Solution.go
package main

import "fmt"

type TreeNode struct {
    Val   int
    Left  *TreeNode
    Right *TreeNode
}

func createTreeUsingBFS(nums []int) *TreeNode {
    arrLen := len(nums)
    var nodeArr []*TreeNode

    for i := 0; i < arrLen; i++ {
        nodeArr = append(nodeArr, &TreeNode{nums[i], nil, nil})
        if i == 0 { continue }
        if i % 2 == 1 {
            nodeArr[(i - 1) / 2].Left = nodeArr[i]
        } else {
            nodeArr[(i - 1) / 2].Right = nodeArr[i]
        }
    }

    return nodeArr[0]
}

func sumNumbers(root *TreeNode) int {
    if root == nil { return 0 }

    return dfs(root, 0)
}

func dfs(root *TreeNode, sum int) int {
    if root == nil { return 0 }
    if root.Left == nil && root.Right == nil { return sum + root.Val }
    sum += root.Val
    return dfs(root.Left, 10 * sum) + dfs(root.Right, 10 * sum)
}

func main() {
    nums := []int{4, 9, 0, 5, 1}
    root := createTreeUsingBFS(nums)
    fmt.Println(sumNumbers(root))
}

File: 129_sum_root_to_leaf_numbers/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {};
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public static TreeNode createTreeUsingBFS(int[] nums) {
        TreeNode[] nodeArr = new TreeNode[nums.length];
        for (int i = 0; i < nums.length; i++) {
            nodeArr[i] = new TreeNode(nums[i]);
            if (i == 0) continue;
            if (i % 2 == 1) nodeArr[(i - 1) / 2].left = nodeArr[i];
            else nodeArr[(i - 1) / 2].right = nodeArr[i];
        }

        return nodeArr[0];
    }

    public int sumNumbers(TreeNode root) {
        return dfs(root, 0);
    }

    public int dfs(TreeNode root, int sum) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return sum + root.val;
        sum += root.val;
        return dfs(root.left, 10 * sum) + dfs(root.right, 10 * sum);
    }

    public static void main(String[] args) {
        int[] nums = {4, 9, 0, 5, 1};
        TreeNode root = createTreeUsingBFS(nums);
        System.out.println(new Solution().sumNumbers(root));
    }
}

